source/common/formatter/substitution_formatter.cc

```diff
@@ -1918,6 +1926,48 @@ UpstreamHostMetadataFormatter::UpstreamHostMetadataFormatter(const std::string&
                           return host->metadata().get();
                         }) {}
 
+// Parses the proxy protocol TLVs stored in the FilterState with the key "envoy.network.proxy_protocol_options" 
+// and returns a function suitable for accessing the specified metadata from an StreamInfo::StreamInfo. 
+absl::optional<std::string> ProxyProtocolTlvsFormatter::format(const Http::RequestHeaderMap&,
+                                                        const Http::ResponseHeaderMap&,
+                                                        const Http::ResponseTrailerMap&,
+                                                        const StreamInfo::StreamInfo& stream_info,
+                                                        absl::string_view) const {
+  const Envoy::StreamInfo::FilterState& filter_state = stream_info.filterState();
+  auto typed_state = filter_state.getDataReadOnly<Network::ProxyProtocolFilterState>(Network::ProxyProtocolFilterState::key());
+
+    // Value exists but isn't string accessible is a contract violation; throw an error.
+    if (typed_state == nullptr) {
+      ENVOY_LOG_MISC(debug, "Invalid header: PROXY_PROTOCOL_TLVS.");
+      return std::string();
+    }
+
+  // Format the vector of proxy protocol TLVs as follows: <type integer>|<base64 encoded value>[,...]
+    std::ostringstream oss;
+  if (!typed_state->value().tlv_vector_.empty()) {
+      ENVOY_LOG_MISC(debug, "parsing proxy protocol TLVs: length of the TLV vector is {}", typed_state->value().tlv_vector_.size());
+
+      // Add the first element with no delimiter
+      auto first_tlv = typed_state->value().tlv_vector_.cbegin();
+
+      // Base64 encode the TLV value
+      auto base64_encoded_tlv_value = Base64::encode(first_tlv->value.data(), first_tlv->value.length());
+      oss << fmt::format("{}|{}", first_tlv->type, base64_encoded_tlv_value);
```

Comment:
If `FilterStates` can be printed out as a JSON, like`[{"type": 5, "value": "xxxx" }]`, that sounds better than a custom format.

Reply:
Sure! I used this format because I thought it was specifically requested [here](https://mongodb.slack.com/archives/C048E90HM5G/p1671486434953279?thread_ts=1671485739.355989&cid=C048E90HM5G), but happy to switch to JSON.

Reply:
I see. If we have to invent a new format, then it makes sense since it's simple. If using FilterState formatter is possible, then I think JSON is already supported and makes more sense. @betsybutton do you mind looking into its feasibility and also asking Dan if JSON format is acceptable?

Reply:
Absolutely! JSON formatting is feasible here. I'll ask Dan what he thinks.

Reply:
JSON is fine if we need to do the output that way, but if we're going with the `%PROXY_PROTOCOL_TLVS(<tlv_type>)%` approach, I think the output can just be the b64 encoded TLV value, which will simplify things for baas and any other consumers in the future.

Reply:
@danhcole Yes, we can just do the base64 encoded value(s) for the specified type. This thread became out-of-date after the other conversations about formatting :) Will push changes soon to implement the formatting described [here](https://github.com/10gen/envoy-serverless/pull/6#discussion_r1069617943).